packages/components/src/confirm-dialog/types.ts

```diff
@@ -0,0 +1,23 @@
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line no-restricted-imports
+import type { SyntheticEvent } from 'react';
+
+interface Props {
+	isOpen?: undefined;
```

Comment:
This type doesn't make much sense to me; we should either change it to

```suggestion
	isOpen?: boolean;
```

or remove this line entitely

Reply:
`isOpen` might be `undefined` though. I explained it [here](https://github.com/WordPress/gutenberg/pull/34153#discussion_r715701219).

Reply:
> isOpen might be undefined though

Yes, but typing it as `isOpen?: undefined;` means that it can _only_ be undefined — either by being omitted (`?` notation) or by being explicitly assigned to `undefined`

Reply:
Ah yeah, I see what you mean. It's indeed weird as it's an optional `undefined` type, which indeed doesn't make much sense. I'll remove the line and TS should still detect the proper type structure in the union.

Reply:
I think I may have found a better way to type these props. The main change would be using the `never` type to indicate that the uncontrolled component should not have the `isOpen` prop defined

```diff
diff --git a/packages/components/src/confirm-dialog/types.ts b/packages/components/src/confirm-dialog/types.ts
index 2a4312ac42..ed0eca8ca8 100644
--- a/packages/components/src/confirm-dialog/types.ts
+++ b/packages/components/src/confirm-dialog/types.ts
@@ -2,25 +2,26 @@
  * External dependencies
  */
 // eslint-disable-next-line no-restricted-imports
-import type { MouseEvent, KeyboardEvent } from 'react';
+import type { MouseEvent, KeyboardEvent, ReactNode } from 'react';
 
 export type DialogInputEvent =
 	| KeyboardEvent< HTMLDivElement >
 	| MouseEvent< HTMLButtonElement >;
 
-interface BaseProps {
+type BaseProps = {
 	title?: string;
-	message: React.ReactNode;
+	message: ReactNode;
 	onConfirm: ( event: DialogInputEvent ) => void;
-	onCancel?: ( event: DialogInputEvent ) => void;
-}
+};
 
-interface ControlledProps extends BaseProps {
+type ControlledProps = BaseProps & {
+	onCancel: ( event: DialogInputEvent ) => void;
 	isOpen: boolean;
-}
+};
 
-interface UncontrolledProps extends BaseProps {
-	isOpen?: undefined;
-}
+type UncontrolledProps = BaseProps & {
+	onCancel?: ( event: DialogInputEvent ) => void;
+	isOpen: never;
+};
 
 export type OwnProps = ControlledProps | UncontrolledProps;
```

Reply:
Oh, this does look better! I'm wondering why a `type` instead of an `interface`, though? I mean, I understand the `type` keyword better communicates what our type will be used for, I guess, but in the end, they would result in the same structural typing for TS, no?

Reply:
I had to make `isOpen` optional in the `UncontrolledProps` type for it to work, though:

```
type UncontrolledProps = BaseProps & {
	onCancel?: ( event: DialogInputEvent ) => void;
	isOpen?: never;
};
```

Reply:
Changes in https://github.com/WordPress/gutenberg/pull/34153/commits/d130ed19905ce2e0c328415bd80fec8c4e532beb.